### PR TITLE
TYP: Add overload to specify output of Colormap.__call__ when possible

### LINKED ISSUE #26501: [Bug]: type-checking errors with mypy + matplotlib 3.8.0rc1
### Bug summary

In testing type-checking integration between matplotlib 3.8.0rc1 and [cmyt](https://github.com/yt-project/cmyt/pull/134), I found mypy flagged a couple errors that I think may be spurious ?

Here's a minimal reproducer of the issue I'm talking about.

### Code for reproduction

```python
# t.py , linted with `mypy t.py`
from matplotlib.colors import Colormap                                                                           
import numpy as np                                                                                               
def rgb_from_cmap(cmap: Colormap) -> np.ndarray:                                                                 
    RGBA = cmap(np.linspace(0, 1, 155))                                                                          
    RGB = RGBA[:, :3]                                                                                            
    return RGB
```

### Actual outcome

```
t.py: note: In function "rgb_from_cmap":
t.py:5: error: No overload variant of "__getitem__" of "tuple" matches argument type "Tuple[slice, slice]"  [call-overload]
t.py:5: note: Possible overload variants:
t.py:5: note:     def __getitem__(self, SupportsIndex, /) -> float
t.py:5: note:     def __getitem__(self, slice, /) -> Tuple[float, ...]
t.py:6: error: Incompatible return value type (got "Union[float, ndarray[Any, Any]]", expected "ndarray[Any, Any]")  [return-value]
Found 2 errors in 1 file (checked 1 source file)
```

### Expected outcome

no error. The second one is easy enough to handle in user code by returning `np.asarray(RGB)` (though I am not 100% sure that this should be necessary), but the first error really feels undesired to me; 

Quoting from `Colormap.__call__`'s docstring:
>        Returns
>        -------
>        Tuple of RGBA values if X is scalar, otherwise an array of
>        RGBA values with a shape of ``X.shape + (4, )``.

it should be clear that the return value is necessarily an array with `ndim >= 2` in my case ?
I think the problem is that the return type of this function is annotated as a union type 
https://github.com/matplotlib/matplotlib/blob/dd091941cb745b6c72ffa4c4deb5c87159bf9c02/lib/matplotlib/colors.pyi#L74

However it's not clear to me how to fix it.

### Additional information

previous versions of matplotlib "type-checked" correctly (because everything was assumed to be `Any` !)

### Operating system

ubuntu

### Matplotlib Version

3.8.0rc1

### Matplotlib Backend

N/A

### Python version

3.11.4

### Jupyter version

N/A

### Installation

pip

### PR DESCRIPTION
**Pull request by ksunden**
## PR summary

Closes #26501

Does much the same as #26434, but for `Colormap.__call__`, noting that the scalar->tuple case is a little different to scalar->scalar

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [n/a] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

**Review by neutrinoceros — APPROVED** [review]
thank you so much for the quick patch ! I confirm this fixes the issue

File: lib/matplotlib/colors.pyi [hunk 1]
```diff
@@ -69,6 +69,15 @@ class Colormap:
...
     colorbar_extend: bool
     def __init__(self, name: str, N: int = ...) -> None: ...
+    @overload
+    def __call__(
+        self, X: Sequence[float] | np.ndarray, alpha: ArrayLike | None = ..., bytes: bool = ...
+    ) -> np.ndarray: ...
+    @overload
+    def __call__(
+        self, X: float, alpha: float | None = ..., bytes: bool = ...
+    ) -> tuple[float, float, float, float]: ...
```
Comment (rcomer) [inline lib/matplotlib/colors.pyi:79]: I’m trying to get my head around why the third option below still needs the union.  Are there inputs other than `X: float, alpha: float` that would return a tuple?  Or does the checker check the `float, float` combination against both because floats are array-like and the checker doesn’t stop after finding the precise match?  Or something else?

Comment (ksunden) [reply lib/matplotlib/colors.pyi:79]: `ArrayLike` is a fair bit broader than I would actually like in many cases. Numpy (understandably) does not provide all of its building blocks as public apis, though, so narrowing it would essentially mean reimplementing or censoring a lot of the machinery. But I suspect there are things that do qualify under arraylike that may not qualify as Sequence (particularly things that provide a `__array__` method, though probably still inadvisable to make such a thing that isn't considered a sequence)

Mypy does not like when multiple signatures are given that are incompatible, though and float is included in arraylike, so it's output must also be in the return type.

One option would be to make the fallback `Any` so that users won't have to deal with a union return type but mypy also doesn't complain. It's a bit of a cop out answer, but sometimes best for situations where the return type is known but cannot be differentiated by input types (in this case because we want the catch all arraylike to be sure we didn't miss anything)


The other aspect I didn't fully think about was if other scalars could be passed, though pretty sure it essentially is floats (and int I guess, but type checkers accept int when typed as float)

TLDR it's a conservative catch all that probably isn't going to be selected in most use cases, but needs to be consistent with the other signature options

**Review by rcomer — APPROVED** [review]

**Review by QuLogic — APPROVED** [review]
